lms/djangoapps/program_enrollments/api/v1/views.py

```diff
@@ -0,0 +1,249 @@
+        if not good_count:
+            return Response(results, HTTP_422_UNPROCESSABLE_ENTITY)
+        if good_count != len(results):
+            return Response(results, HTTP_207_MULTI_STATUS)
+        else:
+            return Response(results)
+
+    def check_enrollment_request(self, enrollment, seen_student_keys):
+        """
+        Checks that the given enrollment record is valid and hasn't been duplicated
+        """
+        student_key = enrollment['student_key']
+        if student_key in seen_student_keys:
+            return CourseEnrollmentResponseStatuses.DUPLICATED
+        seen_student_keys.add(student_key)
+        enrollment_serializer = ProgramCourseEnrollmentRequestSerializer(data=enrollment)
+        try:
+            enrollment_serializer.is_valid(raise_exception=True)
+        except ValidationError as e:
+            if enrollment_serializer.has_invalid_status():
+                return CourseEnrollmentResponseStatuses.INVALID_STATUS
+            else:
+                raise e
+
+    def get_existing_program_enrollments(self, enrollments):
+        """
+        Parameters:
+            - enrollments: A list of enrollment requests
+        Returns:
+            - Dictionary mapping all student keys in the enrollment requests
+              to that user's existing program enrollment <self.program>
+        """
+        external_user_keys = [e["student_key"] for e in enrollments]
+        existing_enrollments = ProgramEnrollment.objects.filter(
+            external_user_key__in=external_user_keys
+        )
+        existing_enrollments = existing_enrollments.prefetch_related('program_course_enrollments')
+        return {enrollment.external_user_key: enrollment for enrollment in existing_enrollments}
+
+    def enroll_learner_in_course(self, enrollment_request, program_enrollments):
+        """
+        Attempts to enroll the specified user into the course as a part of the
+         given program enrollment with the given status
+
+        Returns the actual status
+        """
+        student_key = enrollment_request['student_key']
+        try:
+            program_enrollment = program_enrollments[student_key]
+        except KeyError:
+            return CourseEnrollmentResponseStatuses.NOT_IN_PROGRAM
+        if program_enrollment.get_program_course_enrollment(self.course_key):
+            return CourseEnrollmentResponseStatuses.CONFLICT
+
+        status = enrollment_request['status']
+        course_enrollment = None
+        if program_enrollment.user:  # This user has an account, enroll them in the course
+            course_enrollment = CourseEnrollment.enroll(
+                program_enrollment.user,
+                self.course_key,
+                mode=CourseMode.MASTERS,
+                check_access=True,
+            )
+            if status == CourseEnrollmentResponseStatuses.INACTIVE:
+                course_enrollment.deactivate()
```

Comment:
Are there any potential side-effect of enrolling/unenrolling?  I see we can just create the inactive enrollment record with `get_or_create_enrollment`.  Would that be more appropriate here?

Reply:
`get_or_create_enrollment` doesn't take course mode as an argument, so I thought it would be more straightforward to use  `enroll`.